File: visitor-comparison/src/main/java/com/edwyn/demo/visitor/java25/payment/PaymentCalculator.java
package com.edwyn.demo.visitor.java25.payment;

import com.edwyn.demo.visitor.java25.customer.Customer;
import com.edwyn.demo.visitor.java25.customer.Premium;
import com.edwyn.demo.visitor.java25.customer.Standard;
import com.edwyn.demo.visitor.java25.customer.Vip;

public final class PaymentCalculator {

    private PaymentCalculator() {}

    public static PaymentResult calculate(int totalCents, Customer customer, PaymentMethod payment) {
        return switch (customer) {
            case Premium p -> switch (payment) {
                case CardPayment c    -> result(totalCents, 0,                    "PREMIUM x Carte " + c.network() + " : gratuit");
                case CryptoPayment cr -> result(totalCents,  pct(totalCents, 3),  "PREMIUM x Crypto " + cr.currency() + " : cashback 3%");
            };
            case Vip v -> switch (payment) {
                case CardPayment c                              -> result(totalCents, -pct(totalCents, 1), "VIP x Carte " + c.network() + " : frais 1%");
                case CryptoPayment cr when totalCents > 20_000  -> result(totalCents,  pct(totalCents, 1), "VIP x Crypto " + cr.currency() + " (> 200 EUR) : cashback 1%");
                case CryptoPayment cr                           -> result(totalCents, 0,                   "VIP x Crypto " + cr.currency() + " (<= 200 EUR) : neutre");
            };
            case Standard s -> switch (payment) {
                case CardPayment c    -> result(totalCents, -pct(totalCents, 2), "STANDARD x Carte " + c.network() + " : frais 2%");
                case CryptoPayment cr -> result(totalCents, -pct(totalCents, 1), "STANDARD x Crypto " + cr.currency() + " : frais 1%");
            };
        };
    }

    private static int pct(int amount, int percent) {
        return amount * percent / 100;
    }

    private static PaymentResult result(int base, int adjustment, String description) {
        return new PaymentResult(base, adjustment, base + adjustment, description);
    }
}

File: visitor-comparison/src/main/kotlin/com/edwyn/demo/visitor/kotlin/payment/PaymentCalculator.kt
package com.edwyn.demo.visitor.kotlin.payment

import com.edwyn.demo.visitor.kotlin.customer.Customer
import com.edwyn.demo.visitor.kotlin.customer.Premium
import com.edwyn.demo.visitor.kotlin.customer.Standard
import com.edwyn.demo.visitor.kotlin.customer.Vip

object PaymentCalculator {
    fun calculate(totalCents: Int, customer: Customer, payment: PaymentMethod): PaymentResult =
        when (customer) {
            is Premium -> when (payment) {
                is CardPayment -> result(totalCents, 0, "PREMIUM x Carte ${payment.network} : gratuit")
                is CryptoPayment -> result(totalCents, pct(totalCents, 3), "PREMIUM x Crypto ${payment.currency} : cashback 3%")
            }

            is Vip -> when (payment) {
                is CardPayment -> result(totalCents, -pct(totalCents, 1), "VIP x Carte ${payment.network} : frais 1%")
                is CryptoPayment if totalCents > 20_000 ->
                    result(totalCents, pct(totalCents, 1), "VIP x Crypto ${payment.currency} (> 200 EUR) : cashback 1%")

                is CryptoPayment ->
                    result(totalCents, 0, "VIP x Crypto ${payment.currency} (<= 200 EUR) : neutre")
            }

            is Standard -> when (payment) {
                is CardPayment -> result(totalCents, -pct(totalCents, 2), "STANDARD x Carte ${payment.network} : frais 2%")
                is CryptoPayment -> result(totalCents, -pct(totalCents, 1), "STANDARD x Crypto ${payment.currency} : frais 1%")
            }
        }

    private fun pct(amount: Int, percent: Int): Int =
        amount * percent / 100

    private fun result(base: Int, adjustment: Int, description: String): PaymentResult =
        PaymentResult(base, adjustment, base + adjustment, description)
}
